main.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+)
+
+type Command struct {
+	Name     string
+	Argument string
+}
+
+func readInput() (<-chan Command, []string) {
+	cmds := stdIn()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var filePath string
+	out := make(chan Command, len(cmds))
+
+	go func(cmds []string, filePath *string) {
+		for i, cmd := range cmds {
+			sepCmd := Split(cmd)
+			if i == 0 && len(sepCmd) == 1 {
+				log.Fatal("filepath not given")
+			}
+			command := Command{}
+			for k, name := range sepCmd {
+				if command.Name == "" {
+					command.Name = strings.ToLower(name)
+				} else if name != "" {
+					if *filePath == "" && k == len(sepCmd)-1 {
+						*filePath = name
+						wg.Done()
+						continue
+					} else {
+						command.Argument = name
+					}
+				}
+			}
+			out <- command
+		}
+		close(out)
+	}(cmds, &filePath)
+
+	wg.Wait()
+	fileText := readFile(filePath)
+	return out, fileText
+}
+
+func execute(cmds <-chan Command, input []string) []string {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(cmds <-chan Command) {
+		defer wg.Done()
+		for cmd := range cmds {
+			switch cmd.Name {
+			case "cat":
+				continue
+			case "grep":
+				grepFunc(&input, cmd.Argument)
+			case "sort":
+				sortFunc(input)
+			default:
+				log.Fatal("unknown command")
+			}
+		}
+	}(cmds)
+	wg.Wait()
+	return input
+}
+
+func sortFunc(input []string) {
+	sort.Strings(input)
+}
+
+func grepFunc(input *[]string, arg string) {
+	if arg == "" {
+		log.Fatal("empty arg for grep")
+	}
+	for i := len(*input) - 1; i >= 0; i-- {
```

Comment:
Why loop moves backwards?

Reply:
to be sure, that I reached all the elements of slice